Splay Tree/SplayTree.swift

```diff
@@ -0,0 +1,369 @@
+      return true
+      
+    } else {
+      
+      return false
+    }
+  }
+  
+  func traverseInorder() -> [T] {
+    
+    var traversal: [T] = []
+    
+    recursiveTraverseInorder(root, traversal: &traversal)
+    
+    return traversal
+  }
+  
+  func traversePreorder() -> [T] {
+    
+    var traversal: [T] = []
+    
+    recursiveTraversePreorder(root, traversal: &traversal)
+    
+    return traversal
+  }
+  
+  func traversePostorder() -> [T] {
+    
+    var traversal: [T] = []
+    
+    recursiveTraversePostorder(root, traversal: &traversal)
+    
+    return traversal
+  }
+  
+  func traverseBreadthFirst() -> [T] {
+    return []
+  }
+  
+  //MARK: Private
+  
+  private func findNode(_ value: T) -> SplayTreeNode<T>? {
+    
+    var current: SplayTreeNode<T>? = root
+    var result: SplayTreeNode<T>?
+    
+    while (current != nil) {
+      
+      if current!.value < value {
+        
+        current = current!.right
+        
+      } else if current!.value > value {
+        
+        current = current!.left
+        
+      } else {
+        
+        splay(current!)
+        result = current
+        break
+      }
+    }
+    
+    return result
+  }
+  
+  private func maxInSubtree(_ root: SplayTreeNode<T>) -> SplayTreeNode<T> {
+    
+    var node: SplayTreeNode<T> = root
+    
+    while (node.right != nil) {
+      
+      node = node.right!
+    }
+    
+    return node
+  }
+  
+  private func replace(node x: SplayTreeNode<T>, with y: SplayTreeNode<T>?) {
+    
+    if x.parent == nil {
+      root = y
+    } else if x == x.parent?.left {
+      x.parent?.left = y
+    } else {
+      x.parent?.right = y
+    }
+    
+    if y != nil {
+      y?.parent = x.parent
+    }
+  }
+  
+  private func rotateLeft(_ x: SplayTreeNode<T>) {
+    
+    let y = x.right
+    x.right = y?.left
+    
+    if y?.left != nil {
+      y?.left?.parent = x
+    }
+    
+    y?.parent = x.parent
+    
+    if x.parent == nil {
+      
+      root = y
+      
+    } else if x == x.parent?.left {
+      
+      x.parent?.left = y
+      
+    } else {
+      
+      x.parent?.right = y
+    }
+    
+    y?.left = x
+    x.parent = y
+  }
+  
+  private func rotateRight(_ x: SplayTreeNode<T>) {
+    
+    let y = x.left
+    x.left = y?.right
+    
+    if y?.right != nil {
+      y?.right?.parent = x
+    }
+    
+    y?.parent = x.parent
+    
+    if x.parent == nil {
+      
+      root = y
+      
+    } else if x == x.parent?.left {
+      
+      x.parent?.left = y
+      
+    } else {
+      
+      x.parent?.right = y
+    }
+    
+    y?.right = x
+    x.parent = y
+  }
+  
+  private func splay(_ x: SplayTreeNode<T>) {
+    
+    while x.parent != nil {
+      
+      if x.parent!.parent == nil {
+        
+        if x == x.parent!.left {
+          rotateRight(x.parent!)
+        } else {
+          rotateLeft(x.parent!)
+        }
+        
+      } else if x == x.parent!.left && x.parent == x.parent!.parent!.left {
```

Comment:
Aside from the let binding changes, maybe it would be nice to make the conditions more readable by breaking it up into helper methods? Not a big deal but Splay trees have three rotational operations, such as zig zig, zig zag, and single rotation.